mirgecom/boundary.py

```diff
@@ -575,117 +596,202 @@ def __init__(self):
+
+        grad_species_mass_bc = 1.*grad_cv_minus.species_mass
+        if nspecies > 0:
+            from mirgecom.fluid import species_mass_fraction_gradient
+            grad_y_minus = species_mass_fraction_gradient(state_minus.cv,
+                                                          grad_cv_minus)
+            grad_y_bc = grad_y_minus - np.outer(grad_y_minus@normal, normal)
+            grad_species_mass_bc = 0.*grad_y_bc
+
+            for i in range(nspecies):
+                grad_species_mass_bc[i] = \
+                    (state_minus.mass_density*grad_y_bc[i]
+                     + state_minus.species_mass_fractions[i]*grad_cv_minus.mass)
+
+        return grad_species_mass_bc
+
+
+# Note: callback function inputs have no default on purpose so that they can't be
+# accidentally omitted
+def _inviscid_flux_for_prescribed_state_mengaldo(
+        dcoll, dd_bdry, gas_model, state_minus, *, state_plus_func,
+        numerical_flux_func=inviscid_facial_flux_rusanov,
+        **kwargs):
+    """Compute the inviscid boundary flux."""
+    if state_plus_func is None:
+        state_plus_func = _identical_state
+
+    dd_bdry = as_dofdesc(dd_bdry)
+    normal = state_minus.array_context.thaw(dcoll.normal(dd_bdry))
+
+    state_plus = state_plus_func(
+        dcoll, dd_bdry, gas_model, state_minus, **kwargs)
+    state_pair = TracePair(dd_bdry, interior=state_minus, exterior=state_plus)
+
+    return numerical_flux_func(state_pair, gas_model, normal)
+
+
+# Note: callback function inputs have no default on purpose so that they can't be
+# accidentally omitted
+def _viscous_flux_for_prescribed_state_mengaldo(
+        dcoll, dd_bdry, gas_model, state_minus, grad_cv_minus,
+        grad_t_minus, *, state_bc_func, grad_cv_bc_func, grad_temperature_bc_func,
+        numerical_flux_func=viscous_facial_flux_central,
+        **kwargs):
+    """Compute the viscous boundary flux."""
+    if state_bc_func is None:
+        state_bc_func = _identical_state
+    if grad_cv_bc_func is None:
+        grad_cv_bc_func = _identical_grad_cv
+    if grad_temperature_bc_func is None:
+        grad_temperature_bc_func = _identical_grad_temperature
+
+    dd_bdry = as_dofdesc(dd_bdry)
+    normal = state_minus.array_context.thaw(dcoll.normal(dd_bdry))
+
+    state_bc = state_bc_func(
+        dcoll=dcoll, dd_bdry=dd_bdry, gas_model=gas_model,
+        state_minus=state_minus, **kwargs)
+
+    grad_cv_bc = grad_cv_bc_func(
+        dcoll=dcoll, dd_bdry=dd_bdry, gas_model=gas_model,
+        state_minus=state_minus, state_bc=state_bc, grad_cv_minus=grad_cv_minus,
+        grad_t_minus=grad_t_minus, **kwargs)
+
+    grad_t_bc = grad_temperature_bc_func(
+        dcoll=dcoll, dd_bdry=dd_bdry, gas_model=gas_model,
+        state_minus=state_minus, grad_cv_minus=grad_cv_minus,
+        grad_t_minus=grad_t_minus, **kwargs)
+
+    from mirgecom.viscous import viscous_flux
+    return viscous_flux(state_bc, grad_cv_bc, grad_t_bc) @ normal
+
+
```

Comment:
I had a refactor effort of the MengaldoBCs in #858 , which did something similar.  I merged our two efforts, and hope the result is acceptable to you.